teps/XXXX-workspace-paths.md

```diff
@@ -0,0 +1,900 @@
+workspace. Here's an example from our catalog:
+
+    ```yaml
+    workspaces:
+    - name: source
+      description: A workspace where files will be uploaded from.
+    - name: credentials
+      description: A secret with a service account key to use as GOOGLE_APPLICATION_CREDENTIALS.
+    params:
+    - name: path
+      description: The path to files or directories relative to the source workspace that you'd like to upload.
+      type: string
+    - name: serviceAccountPath
+      description: The path inside the credentials workspace to the GOOGLE_APPLICATION_CREDENTIALS key file.
+      type: string
+    ```
+
+Note above that the `path` and `serviceAccountPath` `params` are providing information specific to the
+`source` and `credentials` workspaces. Two things fall out from this:
+
+- Catalog Task authors clearly see a need to allow the paths into Workspaces to
+be customisable by users of their Tasks.
+- The more workspaces that are needed for a Task, the more params the author
+needs to include to provide this customisable path behavior.
+
+At time of writing there are 30 catalog entries employing this pattern. See the
+[Catalog Examples](#catalog-examples) reference for links to all of those.
+
+2. Workspace descriptions are required to explain both the purpose of the workspace
+_and_ the files that are expected to be received on them. Example from
+[recent `tkn` PR](https://github.com/tektoncd/catalog/pull/499/files#diff-6752c482b505c3f8ffed15a7d7d291b5R20-R23):
+
+> - **kubeconfig**: An [optional workspace](https://github.com/tektoncd/pipeline/blob/master/docs/workspaces.md#using-workspaces-in-tasks)
+> that allows you to provide a `.kube/config` file for `tkn` to access the cluster.
+> The file should be placed at the root of the Workspace with name `kubeconfig`.
+
+### Motivation 4: "Woolly" Reasons
+
+The following items also motivate this TEP but to a lesser degree:
+
+- Decouple declaration of the files and directories a Task needs from their
+eventual "location" inside the Task Steps' containers.
+- Provide a kind of "structural typing" for the contents of a workspace, and a
+way for different but compatible "types" to interface by mapping paths.
+
+### Goals
+
+1. Provide a way for Task authors to declare paths expected on a workspace.
+2. Provide a way for Task authors to declare paths that will be produced on a workspace.
+3. Provide a way for Pipeline authors to declare important files that one PipelineTask produces and another expects.
+4. Allow overriding the paths in TaskRuns and Pipelines/PipelineRuns.
+5. Provide clear errors when expected or created paths are missing.
+6. Provide a blessed way of declaring paths on a workspace in catalog tasks.
+7. Inject new variables that allow Task and Pipeline authors to access paths without hard-coding them.
+
+### Non-Goals
+
+- Declaring any other file features like file types, permissions, content
+validation, etc...
```

Comment:
Maybe this isn't quite the same thing, but what would you think about adding one more optional field to the path element for "digest"?

That would act as an integrity check for when a taskrun knows the contents that should be passed in via the workspace. The main use-case here would be consuming packages/tools fetched over the network, where you know the digest ahead of time. This way you don't have to **trust** the curl/wget task to fetch the right thing.

Reply:
I like this idea. I'm wary of increasing the size of this TEP but have started to sketch a follow-up design for `Workspace Path Validations` that this might fit well with?

```yaml
kind: TaskRun
metadata:
  name: run-wget
workspaces:
- name: output
  emptyDir: {}
  paths:
    produced:
    - name: fetched-file
      path: /data.json
      validations:
        digest: "b6323bca0d4d"
        content-validator: "json"
        max-size: "2MB"
```

The exact format here might be wrong. It might be that we want a more generic "annotations" hashmap or we might want this functionality for additional non-Workspace types as well. So I'd like to avoid adding this to the existing TEP if I can avoid it. Seems like a feature worth exploring in depth on its own.